### ISSUE TITLE
Skip AWS automated resources (not skip entire check)

### ISSUE THREAD

**Issue by evandervecht** [bridgecrewio/checkov#4198]
Version: bridgecrew/checkov:2 
We proxy the container through our private hosted Gitlab from Dockerhub.

We are currently deploying an API and some Lambda's. For the loggroups we have logretention enabled. Under the hood it creates a new Lambda to manage the log retention.

Checkov fails on checks CKV_AWS_111, CKV_AWS_116, CKV_AWS_115, CKV_AWS_117 and I want these checks to be enabled for all Lambda's that are defined in Typescript/CDK, and not for automated generated AWS resources. How can I skip based on Cloudformation resources name (i.e. LogRetention*) ?

Typescript/CDK:
`    const lambdaLogGroup = new logs.LogGroup(this, "Loggroup", {
      logGroupName: "/aws/lambda/somename-lambda",
      retention: logs.RetentionDays.ONE_MONTH,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      encryptionKey: kmsKey
    });
`

resulting in CF snippet:
`"LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a": {
  "id": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a",
  "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a",
  "children": {
    "Code": {
      "id": "Code",
      "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/Code",
      "children": {
        "Stage": {
          "id": "Stage",
          "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/Code/Stage",
          "constructInfo": {
            "fqn": "aws-cdk-lib.AssetStaging",
            "version": "2.53.0"
          }
        },
        "AssetBucket": {
          "id": "AssetBucket",
          "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/Code/AssetBucket",
          "constructInfo": {
            "fqn": "aws-cdk-lib.aws_s3.BucketBase",
            "version": "2.53.0"
          }
        }
      },
      "constructInfo": {
        "fqn": "aws-cdk-lib.aws_s3_assets.Asset",
        "version": "2.53.0"
      }
    },
    "ServiceRole": {
      "id": "ServiceRole",
      "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/ServiceRole",
      "children": {
        "ImportServiceRole": {
          "id": "ImportServiceRole",
          "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/ServiceRole/ImportServiceRole",
          "constructInfo": {
            "fqn": "aws-cdk-lib.Resource",
            "version": "2.53.0"
          }
        },
        "Resource": {
          "id": "Resource",
          "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/ServiceRole/Resource",
          "attributes": {
            "aws:cdk:cloudformation:type": "AWS::IAM::Role",
            "aws:cdk:cloudformation:props": {
              "assumeRolePolicyDocument": {
                "Statement": [
                  {
                    "Action": "sts:AssumeRole",
                    "Effect": "Allow",
                    "Principal": {
                      "Service": "lambda.amazonaws.com"
                    }
                  }
                ],
                "Version": "2012-10-17"
              },
              "managedPolicyArns": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition"
                      },
                      ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
                    ]
                  ]
                }
              ]
            }
          },
          "constructInfo": {
            "fqn": "aws-cdk-lib.aws_iam.CfnRole",
            "version": "2.53.0"
          }
        },
        "DefaultPolicy": {
          "id": "DefaultPolicy",
          "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/ServiceRole/DefaultPolicy",
          "children": {
            "Resource": {
              "id": "Resource",
              "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/ServiceRole/DefaultPolicy/Resource",
              "attributes": {
                "aws:cdk:cloudformation:type": "AWS::IAM::Policy",
                "aws:cdk:cloudformation:props": {
                  "policyDocument": {
                    "Statement": [
                      {
                        "Action": [
                          "logs:PutRetentionPolicy",
                          "logs:DeleteRetentionPolicy"
                        ],
                        "Effect": "Allow",
                        "Resource": "*"
                      }
                    ],
                    "Version": "2012-10-17"
                  },
                  "policyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
                  "roles": [
                    {
                      "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB"
                    }
                  ]
                }
              },
              "constructInfo": {
                "fqn": "aws-cdk-lib.aws_iam.CfnPolicy",
                "version": "2.53.0"
              }
            }
          },
          "constructInfo": {
            "fqn": "aws-cdk-lib.aws_iam.Policy",
            "version": "2.53.0"
          }
        }
      },
      "constructInfo": {
        "fqn": "aws-cdk-lib.aws_iam.Role",
        "version": "2.53.0"
      }
    },
    "Resource": {
      "id": "Resource",
      "path": "somepath/LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/Resource",
      "constructInfo": {
        "fqn": "aws-cdk-lib.CfnResource",
        "version": "2.53.0"
      }
    }
  },
  "constructInfo": {
    "fqn": "constructs.Construct",
    "version": "10.1.175"
  }
},`

**Comment by Knifa** [comment]
Hitting this one myself today so I'd like to re-open this.

**Comment by julesmiddleware** [comment]
Was there an answer to this question that I missed? I want to apply scans from inception, so starting with nothing more than the VPC, however, I cannot get the skips to apply?

```
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';

export class CdkStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const vpc = new ec2.Vpc(this, 'CloakVpc', {
      cidr: '10.0.0.0/16',
    });

    // Add Checkov metadata to the VPC construct to skip CKV_AWS_115 check
    this.node.addMetadata('checkov', {
      skip: [
        {
          id: 'CKV_AWS_115',
          comment: 'This Lambda function is automatically created by CDK and does not need a function-level concurrent execution limit'
        },
        {
          id: 'CKV_AWS_116',
          comment: 'This Lambda function is automatically created by CDK and does not need a function-level concurrent execution limit'
        },
        {
          id: 'CKV_AWS_117',
          comment: 'This Lambda function is automatically created by CDK and does not need a function-level concurrent execution limit'
        }
      ]
    });      
  }
}
```

**Comment by evandervecht** [comment]
Hi @julesmiddleware, yes, we did :D, but we haven't done anything on a VPC level.

Herewith the code that took care of our cases.


file: aspects/ignore-checkov.ts
`import { CfnResource, IAspect } from 'aws-cdk-lib';
import { Construct, IConstruct } from 'constructs';
import { CfnPolicy } from 'aws-cdk-lib/aws-iam';
import { IgnoreCheckov } from '../../utils';
import { CfnFunction, Function as LambdaFunction } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Provider } from 'aws-cdk-lib/custom-resources';

/**
 * Aspect that will add checkov ignore metadata to generated CloudFormation resources that are outside
 * of our control. These are mostly about AWS custom resources that are created by CDK constructs to
 * apply some configuration after CloudFormation stack has run if it cannot be done in CF itself.
 *
 * There aren't really any robust methods to check for resources that are created that we cannot directly
 * control, so the checks are mostly name based.
 *
 * *Try to add a reference to the source of the construct so future-you know where it's coming from.*
 */
export class IgnoreCdkGeneratedCheckovWarningsAspect implements IAspect {
  public visit(construct: IConstruct): void {
    this.ignoreDnsValidatedCertificateWarnings(construct);
    this.ignoreAwsLogRetentionWarnings(construct);
    this.ignoreAwsCustomResourceWarnings(construct);
    this.ignoreCustomResourceProviderWarnings(construct);
  }

  private ignoreDnsValidatedCertificateWarnings(construct: IConstruct): void {
    // see https://github.com/aws/aws-cdk/blob/36d356d0b3e422f7451f4b0dd2f971aa0378210e/packages/%40aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts#L99
    const pathSearchString = '/CertificateRequestorFunction/';

    // (1) CKV_AWS_111, policy with write access without constraint
    if (construct instanceof CfnPolicy && construct.node.path.includes(pathSearchString)) {
      IgnoreCheckov(construct, {
        id: 'CKV_AWS_111',
        comment: 'Ignoring check on CDK generated resource (DnsValidatedCertificate)'
      });
    }

    // (2) CKV_AWS_116, DLQ should be configured for a lambda
    if (construct instanceof CfnFunction && construct.node.path.includes(pathSearchString)) {
      IgnoreCheckov(construct, {
        id: 'CKV_AWS_116',
        comment: 'Ignoring check on CDK generated resource (DnsValidatedCertificate)'
      });
    }
  }

  private ignoreAwsLogRetentionWarnings(construct: IConstruct): void {
    // https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-logs/lib/log-retention.ts#L129
    // AWS custom resources will never have a DLQ, don't bother us about it. The resource name will always be
    // of the form 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a as per the linked sources.
    const policySearchString = 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/ServiceRole/DefaultPolicy/Resource';
    if (construct instanceof CfnResource && construct.node.path.includes(policySearchString)) {
      IgnoreCheckov(construct, {
        id: 'CKV_AWS_111',
        comment: 'Ignoring check Ensure IAM policies does not allow write access without constraint'
      });
    }

    const lambdaSearchString = 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a/Resource';
    if (construct instanceof CfnResource && construct.node.path.includes(lambdaSearchString)) {
      IgnoreCheckov(
        construct,
        {
          id: 'CKV_AWS_115',
          comment:
            'Ignoring Ensure that AWS Lambda function is configured for function-level concurrent execution limit'
        },
        {
          id: 'CKV_AWS_116',
          comment: 'Ignoring Ensure AWS Lambda function is configured for a DLQ'
        },
        {
          id: 'CKV_AWS_117',
          comment: 'Ignoring Ensure that AWS Lambda function is configured inside a VPC'
        }
      );
    }
  }

  private ignoreAwsCustomResourceWarnings(construct: IConstruct): void {
    // see https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts#L376
    // AWS custom resources will never have a DLQ, don't bother us about it. The resource name will always be
    // of the form 'AWS<uuid-without-dashes><cdk-id> as per the linked sources.
    const customResourcePattern = /^AWS[0-9a-fA-F]{32}/;

    if (construct instanceof LambdaFunction && construct.node.id.match(customResourcePattern)) {
      IgnoreCheckov(
        construct.node.defaultChild as CfnResource,
        {
          id: 'CKV_AWS_115',
          comment:
            'Ignoring Ensure that AWS Lambda function is configured for function-level concurrent execution limit'
        },
        {
          id: 'CKV_AWS_116',
          comment: 'Ignoring Ensure AWS Lambda function is configured for a DLQ'
        },
        {
          id: 'CKV_AWS_117',
          comment: 'Ignoring Ensure that AWS Lambda function is configured inside a VPC'
        }
      );
    }
  }

  /**
   * When creating custom resources, a provider is also needed which handles the CloudFormation events and sends back a response.
   * The provider creates a lambda function to handle the events. So we have no control on it.
   */
  private ignoreCustomResourceProviderWarnings(construct: IConstruct): void {
    const isCustomResourceProvider = (
      construct: IConstruct
    ): construct is Omit<Provider, 'entrypoint'> & { entrypoint: NodejsFunction } =>
      Construct.isConstruct(construct) && construct instanceof Provider && 'entrypoint' in construct;

    if (isCustomResourceProvider(construct)) {
      const entrypointDefaultChild = construct.entrypoint.node.defaultChild;

      if (Construct.isConstruct(entrypointDefaultChild) && CfnResource.isCfnResource(entrypointDefaultChild)) {
        IgnoreCheckov(entrypointDefaultChild, {
          id: 'CKV_AWS_116',
          comment: 'Ignoring Ensure AWS Lambda function is configured for a DLQ'
        });
      }
    }
  }
}


`

file: utils/ignore-checkov.ts
`import * as cdk from 'aws-cdk-lib';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as s3 from 'aws-cdk-lib/aws-s3';

export function IgnoreCheckov(cfnResource: cdk.CfnResource, ...checks: { id: string; comment?: string }[]) {
  const skipMetadata = checks.map(({ id, comment }) => ({ id, comment: comment ?? `Ignoring checkov check ${id}` }));

  cfnResource.cfnOptions.metadata = {
    ...cfnResource.cfnOptions.metadata,
    checkov: { skip: skipMetadata }
  };
}

export function SkipCheckOvDefaultLambda(lambda: lambda.Function) {
  IgnoreCheckov(
    lambda.node.defaultChild as cdk.aws_lambda.CfnFunction,
    {
      id: 'CKV_AWS_115',
      comment: 'Ensure that AWS Lambda function is configured for function-level concurrent execution limit'
    },
    {
      id: 'CKV_AWS_116',
      comment: 'Ensure that AWS Lambda function is configured for a Dead Letter Queue(DLQ)'
    },
    {
      id: 'CKV_AWS_117',
      comment: 'Ensure that AWS Lambda function is configured inside a VPC'
    },
    {
      id: 'CKV_AWS_173',
      comment: 'Check encryption settings for Lambda environmental variable'
    }
  );
}

export function SkipCheckOvPITRCheck(dynamoTable: dynamodb.Table) {
  IgnoreCheckov(dynamoTable.node.defaultChild as dynamodb.CfnTable, {
    id: 'CKV_AWS_28',
    comment: 'Ensure Dynamodb point in time recovery (backup) is enabled'
  });
}

export function SkipCheckOvDefaultS3(bucket: s3.Bucket) {
  IgnoreCheckov(
    bucket.node.defaultChild as s3.CfnBucket,
    {
      id: 'CKV_AWS_18',
      comment: 'Ensure the S3 bucket has access logging enabled'
    },
    {
      id: 'CKV_AWS_21',
      comment: 'Ensure the S3 bucket has versioning enabled'
    }
  );
}
`

Basically, you have to do this on a higher level L1 construct rather than the L2 constructs.

Hope this helps.

**Comment by julesmiddleware** [comment]
Thank you, I would never have worked this out alone. I'm working out how to get it to match my patterns, but thanks again for the structure to get me started.
